File: frontend/src/engine/capabilities/highlightDuplicates.ts
/**
 * highlightDuplicates — add a conditional-formatting rule that colors
 * duplicate values in a range.
 *
 * One-step equivalent of addConditionalFormat with a COUNTIF formula. Uses
 * a duplicateValues rule on ExcelApi 1.6+; falls back to a COUNTIF-based
 * custom formula on older versions.
 */

import {
  CapabilityMeta,
  HighlightDuplicatesParams,
  StepResult,
  ExecutionOptions,
} from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange } from "./rangeUtils";

const meta: CapabilityMeta = {
  action: "highlightDuplicates",
  description: "Highlight duplicate values in a range with a conditional-formatting rule",
  mutates: false,
  affectsFormatting: true,
  requiresApiSet: "ExcelApi 1.6",
};

async function handler(
  context: Excel.RequestContext,
  params: HighlightDuplicatesParams,
  options: ExecutionOptions,
): Promise<StepResult> {
  const { range: address, fillColor = "#FFCCCC", fontColor = "#C50F1F" } = params;

  if (options.dryRun) {
    return { stepId: "", status: "success", message: `Would highlight duplicates in ${address}` };
  }

  options.onProgress?.("Highlighting duplicates...");

  const range = resolveRange(context, address);
  range.load("address");
  await context.sync();

  // Parse the resolved address into an absolute range (for the COUNTIF's
  // FIRST argument — must stay put as the CF is applied to each cell) and a
  // relative cell ref (for the SECOND argument — must shift so the formula
  // evaluates against each cell's OWN value).
  //
  // Previous bug: we used `addr` directly (relative) so when the CF was
  // applied to the second cell, COUNTIF's range shifted by one row and
  // ended up counting values that weren't really duplicates.
  const addr = range.address;
  const sheetPrefix = addr.includes("!") ? addr.substring(0, addr.lastIndexOf("!") + 1) : "";
  const addrOnly = addr.includes("!") ? addr.substring(addr.lastIndexOf("!") + 1) : addr;

  // Build absolute range: "$A$1:$A$100"
  const absoluteRange = addrOnly.split(":").map((part) => {
    const m = part.match(/^\$?([A-Z]+)\$?(\d+)?$/);
    if (!m) return part;
    return `$${m[1]}${m[2] ? `$${m[2]}` : ""}`;
  }).join(":");
  const absRangeRef = `${sheetPrefix}${absoluteRange}`;

  // First cell as a RELATIVE reference so it shifts with each CF application.
  const firstCellMatch = addrOnly.split(":")[0].match(/^\$?([A-Z]+)\$?(\d+)/);
  const firstCell = firstCellMatch ? `${firstCellMatch[1]}${firstCellMatch[2]}` : "A1";

  // The Office.js `ConditionalFormatType` enum in older @types/office-js
  // doesn't expose `duplicateValues`, so we use the `custom` type with a
  // COUNTIF formula — works on ExcelApi 1.6+ and is equivalent visually.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const cf: any = range.conditionalFormats.add(Excel.ConditionalFormatType.custom);
  cf.custom.rule.formula = `=COUNTIF(${absRangeRef},${firstCell})>1`;
  cf.custom.format.fill.color = fillColor;
  cf.custom.format.font.color = fontColor;
  await context.sync();

  return {
    stepId: "",
    status: "success",
    message: `Added duplicate-highlighting rule to ${address}.`,
    outputs: { range: address },
  };
}

registry.register(meta, handler as any);
export { meta };

File: backend/app/execution/capabilities/highlight_duplicates.py
"""highlightDuplicates — conditional-format duplicate values in a range.

Uses the COM FormatConditions collection to add a duplicate-values rule
directly. On platforms without the DuplicateValues enum, falls back to a
COUNTIF-based custom formula.
"""

from __future__ import annotations

from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range
from app.execution.capabilities.tab_color import _hex_to_bgr_int  # reuse


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    address = params.get("range")
    fill_color = params.get("fillColor") or "#FFCCCC"
    font_color = params.get("fontColor") or "#C50F1F"
    if not address:
        return {"status": "error", "message": "highlightDuplicates requires 'range'."}

    rng = resolve_range(ctx.workbook_handle, address)

    if ctx.dry_run:
        return {"status": "preview", "message": f"Would highlight duplicates in {rng.address}."}

    try:
        # Try the DuplicateValues CF type first (COM xlDuplicate = 8,
        # AddTop10 enum int). Excel COM: FormatConditions.AddUniqueValues(),
        # then set DupeUnique = 0 (xlDuplicate) / 1 (xlUnique).
        cf = rng.api.FormatConditions.AddUniqueValues()
        cf.DupeUnique = 0  # xlDuplicate
        cf.Interior.Color = _hex_to_bgr_int(fill_color)
        cf.Font.Color = _hex_to_bgr_int(font_color)
    except Exception:
        # Fallback to custom formula.
        try:
            addr = rng.address
            # Use the top-left of the range as the anchor cell in COUNTIF.
            first_cell = rng[0, 0].address.replace("$", "")
            cf = rng.api.FormatConditions.Add(
                Type=2,  # xlExpression
                Formula1=f"=COUNTIF({addr},{first_cell})>1",
            )
            cf.Interior.Color = _hex_to_bgr_int(fill_color)
            cf.Font.Color = _hex_to_bgr_int(font_color)
        except Exception as exc:  # noqa: BLE001
            return {
                "status": "error",
                "message": f"Could not add duplicate-highlight rule: {exc}",
                "error": str(exc),
            }

    return {
        "status": "success",
        "message": f"Added duplicate-highlight rule on {rng.address}.",
        "outputs": {"range": rng.address},
    }


registry.register("highlightDuplicates", handler, mutates=False, affects_formatting=True)
